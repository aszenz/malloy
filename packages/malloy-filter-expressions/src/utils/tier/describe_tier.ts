/**
 * Copyright (c) 2023 Google LLC
 * SPDX-License-Identifier: MIT
 */
import defaultTo from 'lodash/defaultTo';
import keyBy from 'lodash/keyBy';
import {FilterItemToStringMapType, FilterModel} from '../..';
import {FilterExpressionType} from '../../types';
import {addQuotes} from '../string/add_quotes';
import {describeIsItem} from '../summary/describe_is_item';
import {describeIsAnyValue} from '../summary/describe_is_any_value';
import {joinOr} from '../summary/join_or';
import {escapeParameterValue} from './escape_parameter_value';

const describeMultiValue = (values: string[]) => {
  if (values) {
    return joinOr(values.map(addQuotes));
  }
  return '';
};

const match = ({is, value}: FilterModel, _?: string) => {
  return value && value.length
    ? describeIsItem(is, describeMultiValue(value))
    : describeIsAnyValue();
};

const filterToStringMap: FilterItemToStringMapType = {
  match,
  'anyvalue': describeIsAnyValue,
};

/**
 * Maps a FilterItem to a function for converting it to a filter summary
 */
export const describeTier = (
  item: FilterModel,
  filterType?: FilterExpressionType
): string =>
  defaultTo(filterToStringMap[item.type], () => '')(item, filterType);
