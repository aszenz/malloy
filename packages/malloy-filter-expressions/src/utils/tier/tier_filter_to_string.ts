/**
 * Copyright (c) 2023 Google LLC
 * SPDX-License-Identifier: MIT
 */
import flow from 'lodash/fp/flow';
import {
  FilterASTNode,
  FilterExpressionType,
  FilterItemToStringFunction,
  FilterItemToStringMapType,
  FilterModel,
} from '../..';
import {quoteFilter} from '../string/quote_filter';
import isItemToString from '../to_string/is_item_to_string';
import {escapeParameterValue} from './escape_parameter_value';
import {treeToList} from '../tree/tree_to_list';

const matchToString = ({value, is}: FilterModel, _?: string) => {
  return (
    isItemToString(is, '', '-') +
    value
      .map((val: string) => quoteFilter(val))
      .join(`,${isItemToString(is, '', '-')}`)
  );
};

const anyvalueToString = () => '';

const filterToStringMap: FilterItemToStringMapType = {
  'anyvalue': anyvalueToString,
  'match': matchToString,
};

const serializeTierItem =
  (type: FilterExpressionType) =>
  (item: FilterModel): string => {
    const toStringFunction: FilterItemToStringFunction =
      filterToStringMap[item.type];
    return toStringFunction?.(item, type) || '';
  };
/**
 * Maps a FilterItem to a function for converting it to an expression
 */

const listToExpression =
  (type: FilterExpressionType) => (items: FilterModel[]) =>
    items.map(serializeTierItem(type)).join(',');

/**
 * Converts the AST to an array of FilterItems and then
 * converts each item into its expression representation
 */
export const tierFilterToString = (
  root: FilterASTNode,
  type: FilterExpressionType
): string => flow(treeToList, listToExpression(type))(root);
