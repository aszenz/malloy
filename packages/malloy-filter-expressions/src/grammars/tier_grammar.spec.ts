/**
 * Copyright (c) 2023 Google LLC
 * SPDX-License-Identifier: MIT
 */
import {GrammarTestItem} from '.';
import {
  convertTypeToOption,
  parseFilterExpression,
  summary,
  treeToList,
} from '../utils';
import {tierFilterToString} from '../utils/tier/tier_filter_to_string';
import {tierGrammarTestItems} from './tier_grammar_test_expressions';

const filterType = 'tier';
const testStringItem = (testItem: GrammarTestItem) => {
  test(`${testItem['expression']}`, () => {
    const {expression, type, describe, output} = testItem;
    const ast = parseFilterExpression(filterType, expression);
    const description = summary({type: filterType, expression});
    expect(ast).toMatchSnapshot();
    const list = treeToList(ast);
    const item = list[0];

    const itemType =
      type === 'matchesAdvanced' ? item.type : convertTypeToOption(item);
    expect(itemType).toEqual(type);

    // test descriptions
    expect(description).toBe(describe);

    // test output
    const stringOutput =
      type === 'matchesAdvanced'
        ? expression
        : tierFilterToString(ast, filterType);
    expect(stringOutput).toBe(output);
  });
};

describe('Tier grammar can parse', () => {
  tierGrammarTestItems['forEach'](testStringItem);
});
