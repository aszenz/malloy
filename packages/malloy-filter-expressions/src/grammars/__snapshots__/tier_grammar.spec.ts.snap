// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tier grammar can parse 20 to 29 1`] = `
{
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": [
    "20 to 29",
  ],
}
`;

exports[`Tier grammar can parse sub^_region 1`] = `
{
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": [
    "sub_region",
  ],
}
`;
